### PR TITLE
[SPARK-50018][SQL] Make AbstractStringType serializable

### PR DESCRIPTION
**Pull request by uros-db**
### What changes were proposed in this pull request?
Make `AbstractStringType` class serializable, so that the derived classes can be used in expressions that perform replacement using `Invoke`-like Spark expressions.

### Why are the changes needed?
Objects with custom parameters cannot be used as inputTypes unless the underlying class is serializable. For example, `ValidateUTF8` is a string function that uses `StaticInvoke` replacement, and an object derived from `AbstractStringType` as one of the input types.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing tests suffice. More will be added with appropriate collation support in various Spark expressions.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by uros-db — COMMENTED** [review]
@jovanpavl-db please review, given that you encountered the issue recently

before this is merged, you can manually include the fix in your PR https://github.com/apache/spark/pull/48521 to unblock string function support for collations with trimming sensitivity

**Review by HyukjinKwon — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @uros-db and @HyukjinKwon for review.

**Pull request closed without merging**
